### PR TITLE
Refactor moving of modal elements to solve multiple issues

### LINKED ISSUE #1517: [BUG] Title and footer do not render correctly in modal with outlet in some cases
**Describe the bug**
`<kirby-page-title>` and `<kirby-modal-footer>` render incorrectly when these conditions are met:
- used within modal with outlet
- `<kirby-page-title>` and `<kirby-modal-footer>` are wrapped in sub components and conditionally added to the template i.e. using `*ngIf`
- when navigating to the component using `modalController.navigateWithinModal( ... )`.

**To Reproduce**
Access DRB with mock data on branch `T052/modal-title-footer`

If accessing DRB in a web browser:
1. Go directly to `drb/investment/x-investment/onboarding/(modal:select-profile)` in the URL
2. Click 'Vælg' and then '>'
3. See error
4. Refresh the page and see the modal render correctly.
5. Click '<', 'Vælg' and '>'
6. See the error again

In the app
1. Toggle on 'Jyske Investering' and 'Jyske Investering - Onboarding'
2. Go to 'Investering'
3. Click on 'Start onboarding'
4. Click on 'Start nu > '
5. Check 'Jeg accepterer....'
6. Click 'Lav din plan >'
7. Fill in values and click '>' until arriving at the 'Gennemgå din plan' step
8. See error

**Kirby version**
- 1.4.44

**Expected behavior**
Screenshot 2021-04-28 at 15 08 58

**Screenshots**
Screenshot 2021-04-28 at 15 06 51

**Please complete the following information:**
- OS: [e.g. Windows, MacOS, iOS]
- Browser [e.g. Chrome, Safari]
- Browser version [e.g. 22]

**Additional context**

## Tasks:
- [ ] Have the bug report verified by contacting a member of @kirbydesign/kirby-guild
- [ ] Move issue from "Ready to do" -> "in progress" on our [project board](https://github.com/kirbydesign/designsystem/projects/1)
- [ ] Create Fix Branch
- [ ] Create test reproducing the error
- [ ] Fix
- [ ] If relevant [Update Examples and Showcase](https://cookbook.kirby.design/home/showcase/button)
- [ ] Push commits to GitHub
- [ ] Open a pull request (remember to add `closes #issueno` to the description) and ask @kirbydesign/kirby-guild for review
- [ ] Merge to master and delete branch

### LINKED ISSUE #2096: [Housekeeping] Make getEmbeddedComponentElement independent of markup order
- [X] I have written a [good issue](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Issue) 

## Describe the housekeeping request
As part of solving #843 i've attempted to refactor the `ModalWrapperComponent` method `getEmbeddedComponentElement`. The point of the attempted refactor was to make the function independent of the order of elements in the markup. 

It's very easy to accidentally break the function by introducing new children to `ion-content` - which i had to do in #843. Example of markup that would break the function: 
```html
...
<ion-content [scrollEvents]="true">
  <h1>This heading element breaks getEmbeddedComponentElement</h1>
  <ng-container
    *ngComponentOutlet="config.component; injector: componentPropsInjector"
  ></ng-container>
  <router-outlet
    name="kirbyModalWrapperOutlet"
    [style.visibility]="config.modalRoute ? 'visible' : 'hidden'"
  ></router-outlet>
</ion-content>
...
``` 

After playing around with it for a couple of hours i gave up to move on with the implementation of #843 - i instead changed the function to work with the new markup: 
```typescript 
  private getEmbeddedComponentElement() {
    const { lastElementChild } = this.ionContentElement.nativeElement;
    return !!this.config.modalRoute ? lastElementChild : lastElementChild.previousElementSibling;
  }
```

### LINKED ISSUE #2116: [BUG] kirby-modal-footer does not render correctly when footer is dynamically added and removed
In kirby-modal-wrapper, a MutationObserver is created in order to detect if a kirby-modal-footer is inserted into the modal and move it to the correct place if this is the case.

Unfortunately, this does not work as desired for two reasons:
 - The observer does not observe the entire subtree, and the footer may be nested inside other components
 -  When an element is added containing the kirby-modal-footer it is not detected, because mutation events is not is emitted for elements in the subtree.

The following patch fixes the problem:

```patch
--- a/libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.ts
+++ b/libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.ts
@@ -481,21 +481,27 @@ export class ModalWrapperComponent implements Modal, AfterViewInit, OnInit, OnDe
     const parentElement = this.getEmbeddedComponentElement();
     this.mutationObserver.observe(parentElement, {
       childList: true, // Listen for addition or removal of child nodes
+      subtree: true,
     });
   }
 
   private createEmbeddedElementsMutationObserver(): MutationObserver {
     const observedElements = Object.keys(this.elementToParentMap);
     const callback = (mutations: MutationRecord[]) => {
-      const addedNodes = mutations
-        .filter((mutation) => mutation.type === 'childList') // Filter for mutation to the tree of nodes
-        .map((mutation) => {
-          // Only check for addedNodes as removal is handled by the Angular renderer:
-          return Array.from(mutation.addedNodes).filter((node) =>
-            observedElements.includes(node.nodeName)
-          );
-        });
-
+      const addedNodes: HTMLElement[] = [];
+      for (const mutation of mutations) {
+        for (const element of Array.from(mutation.addedNodes)) {
+          if (element instanceof HTMLElement) {
+            for (const observedElementName of observedElements) {
+              for (const added of Array.from(element.querySelectorAll(observedElementName))) {
+                if (added instanceof HTMLElement) {
+                  addedNodes.push(added);
+                }
+              }
+            }
+          }
+        }
+      }
```

### PR DESCRIPTION
**Pull request by MadsBuchmann**
## Which issue does this PR close?

This PR closes #2096
This PR closes #2116 
This PR closes #1517 

## What is the new behavior?

I'm messing around with refactoring the modal to make it easier to grasp and solve a few bugs. 

### Before 
A mutation observer was used for watching for changes in the embedded component. If a `kirby-page-progress`, `kirby-page-title` or `kirby-modal-footer` element was added, it would be moved to its desired spot. 

The implementation of this was quite complex in my opinion and hard to grasp (And i have a suspiscion that i'm not the only one who believes this 👀)

Furthermore the bug from #2116 stems from the mutation observer not reacting when one of the aforementioned elements are added when they are wrapped in a component defined by the consumer. 

To solve this we could either continue making it more complex (see the first suggested solution on #2116).. Or rethink how the moving of embedded components are done. 

Resulting in ... 

### After 
A new `ModalElementsAdvertiser`! 

What we really want to know is: "which element is added/removed when?" such that we can move them out of the modal content and into the header/footer/where ever it belongs. 

Instead of having the `ModalWrapperComponent` watch for changes, i've switched it around such that the embedded components announce themselves. 

~~Each element that the modal should move registers itself using the method `ModalElementsAdvertiser.registerElement` on `ngAfterViewInit`.~~ 

~~This will in turn trigger an `ModalElementsAdvertiser.[elementType].added$` observable, which the `ModalWrapperComponent` is subscribed to. The `ModalWrapperComponent` can then react accordingly.~~

~~It's more or less a simple pub/sub pattern.~~ 

~~Symmetrically there also exists a `removed$` observable for when `OnDestroy` is triggered for a component.~~

~~It's a bit more verbose but i think it is easier to grasp ones head around as it is very clear what is executed when and has also made it possible to bundle relevant code some more.~~

Edit: new approach! 

We inject the `ModalWrapperComponent` as a `ModalElementsAdvertiser` into the child elements that should be moved. They can then call a `modalElementsAdvertiser.add[ElementName]` and `modalElementsAdvertiser.remove[elementName]` method to register themselves in the wrapper component. 

E.g. when a `kirby-page-title` element is added it calls `modalElementsAdvertiser.addTitle` on ngAfterviewInit and `modalElementsAdvertiser.removeTitle` on ngDestroy.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No (hopefully)

## Are there any additional context?
There's some remaining todos... 

### TODO: 
- [x] The collapse title functionality has decided to not work... I should fix that. 
- [x] ... Related to the above... Some tests are failing. Figure out why. 
- [x] A lot of the tests are actually integration tests. We should split it up such that this is visible. 

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/stable/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [ ] Request that the changes are code-reviewed 
- [ ] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/stable/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be automatically merged to `stable` via [automerge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request).

### CODE REVIEW

File: libs/designsystem/src/lib/components/page/page.component.ts [hunk 1]
```diff
@@ -199,7 +236,7 @@ export class PageComponent
   hasActionsInPage: boolean;
   toolbarTitleVisible: boolean;
   toolbarFixedActionsVisible: boolean;
-  toolbarStickyActionsVisible: boolean;
+  /* TODO: move this logic to the modalWrapper? */ toolbarStickyActionsVisible: boolean;
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/page/page.component.ts:239]: This seems like a runaway comment

File: libs/designsystem/src/lib/components/modal/modal-wrapper/modal-elements-advertiser.service.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  public deregisterElement(
+    elementType: ModalAdvertiserElementType,
+    elementRef: ElementRef<HTMLElement>
+  ) {
+    this.modalElementEntrySubject.next({ type: elementType, action: 'deregister', elementRef });
+  }
+
+  private createModalElementObservableSet(
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/modal/modal-wrapper/modal-elements-advertiser.service.ts:40]:
```suggestion
  private createObservableSet(
```

File: libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.ts [hunk 3]
```diff
@@ -140,6 +142,93 @@ export class ModalWrapperComponent implements Modal, AfterViewInit, OnInit, OnDe
...
+    modalWrapperElement.appendChild(footerElement);
+
+    /* TODO: Is it possible to handle this another way? */
+    this.resizeObserverService.observe(footerElement, (entry) => {
+      const [property, pixelValue] = [
+        '--footer-height',
+        `${Math.floor(entry.contentRect.height)}px`,
+      ];
+      this.setCssVar(this.elementRef.nativeElement, property, pixelValue);
+    });
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.ts:181]: Either spin this off to a new issue or have a chat with @jkaltoft

File: libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.ts [hunk 4]
```diff
@@ -438,111 +470,30 @@ export class ModalWrapperComponent implements Modal, AfterViewInit, OnInit, OnDe
...
-        this.resizeObserverService.observe(childElement, (entry) => {
-          const [property, pixelValue] = [
-            '--footer-height',
-            `${Math.floor(entry.contentRect.height)}px`,
-          ];
-          this.setCssVar(this.elementRef.nativeElement, property, pixelValue);
-        });
-      }
-    });
+    this.renderer.appendChild(newParent, child);
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.ts:484]: To simplify the method it has been changed to only accomodate a single new parent (check type signature on line 479)

File: libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.integration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,368 @@
+import { RouterTestingModule } from '@angular/router/testing';
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.integration.spec.ts:1]: These tests has been seperated out from `modal-wrapper.component.spec.ts` as they're now integration tests with the new implementation.

Comment (mictro) [reply libs/designsystem/src/lib/components/modal/modal-wrapper/modal-wrapper.component.integration.spec.ts:1]: 👍

**Review by mictro — APPROVED** [review]
Very nice result - consider for next show-and-tell 🥇

**Comment by mictro** [comment]
smooth

**Comment by mictro** [comment]
ezgif-2-5661164cfe

**Review by mictro — APPROVED** [review]
Really Nice 👍

**Pull request merged**